src/dnmd/entry.c

```diff
@@ -478,3 +478,374 @@ void free_mdmem(mdcxt_t* cxt, void* mem)
+    save_size += align_to((uint32_t)strlen(heap_name) + 1, 4);
+    // Add the size of the stream itself.
+    // It's not placed directly after the header in the image,
+    // but we might as well account for it here while we're checking
+    // the heap's existence.
+    save_size += heap_size;
+
+    return save_size;
+}
+
+static size_t get_table_stream_size(mdcxt_t* cxt)
+{
+    // II.24.2.6 #~ stream
+    size_t const table_stream_header_size =        
+        + sizeof(uint32_t) // Reserved
+        + sizeof(uint8_t) // MajorVersion
+        + sizeof(uint8_t) // MinorVersion
+        + sizeof(uint8_t) // HeapSizes
+        + sizeof(uint8_t) // Reserved
+        + sizeof(uint64_t) // Valid
+        + sizeof(uint64_t) // Sorted
+        // Rows and Tables entries are both variable length and calculated below
+    ;
+    
+    size_t save_size = table_stream_header_size;
+    
+    for (uint8_t i = 0; i < MDTABLE_MAX_COUNT; ++i)
+    {
+        if (cxt->tables[i].cxt != NULL && cxt->tables[i].row_count != 0)
+        {
+            save_size += sizeof(uint32_t); // Row count
+            save_size += cxt->tables[i].data.size; // Table data
+        }
+    }
+    
+    return save_size;
+}
+
+size_t md_get_save_size(mdhandle_t handle)
+{
+    mdcxt_t* cxt = extract_mdcxt(handle);
+    if (cxt == NULL)
+        return 0;
+    
+    if (cxt->editor == NULL)
+        return cxt->raw_metadata.size;
+    
+    // II.24.2.1 Metadata Root size
+    size_t const image_header_size =
+        sizeof(uint32_t) // Signature
+        + sizeof(uint16_t) // MajorVersion
+        + sizeof(uint16_t) // MinorVersion
+        + sizeof(uint32_t) // Reserved
+        + sizeof(uint32_t) // Length (of version string)
+        + align_to((uint32_t)strlen(cxt->version) + 1, 4) // Version String
+        + sizeof(uint16_t) // Flags
+        + sizeof(uint16_t) // Streams (number of streams)
+    ;
+
+    
+    size_t save_size = image_header_size;
+
+    if (cxt->blob_heap.size != 0)
+        save_size += get_stream_header_and_contents_size("#Blob", cxt->blob_heap.size);
+    if (cxt->guid_heap.size != 0)
+        save_size += get_stream_header_and_contents_size("#GUID", cxt->guid_heap.size);
+    if (cxt->strings_heap.size != 0)
+        save_size += get_stream_header_and_contents_size("#Strings", align_to((uint32_t)cxt->strings_heap.size, 4));
+    if (cxt->user_string_heap.size != 0)
+        save_size += get_stream_header_and_contents_size("#US", cxt->user_string_heap.size);
+
+    if (cxt->context_flags & mdc_minimal_delta)
+        save_size += get_stream_header_and_contents_size("#JTD", 0);
+    
+    // All names of the tables stream are the same length,
+    // so pick the one in the standard.
+    save_size += get_stream_header_and_contents_size("#~", get_table_stream_size(cxt));
+
+    return save_size;
+}
+
+static bool advance_output_stream(uint8_t** data, size_t* data_len, size_t b)
+{
+    return advance_stream((uint8_t const**)data, data_len, b);
```

Comment:
Is the cast needed?

Reply:
Yes, clang complains otherwise that the pointer types are incompatible (with a warning that we have set to turn into an error).